File: python/pyautd3/link/nop.py
"""
File: nop.py
Project: link
Created Date: 10/10/2023
Author: Shun Suzuki
-----
Last Modified: 10/10/2023
Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
-----
Copyright (c) 2023 Shun Suzuki. All rights reserved.

"""


from pyautd3.internal.link import Link, LinkBuilder
from pyautd3.native_methods.autd3capi import (
    NativeMethods as LinkNop,
)
from pyautd3.native_methods.autd3capi_def import ControllerPtr, LinkBuilderPtr, LinkPtr


class Nop(Link):
    """Link which do nothing."""

    class _Builder(LinkBuilder):
        def __init__(self: "Nop._Builder") -> None:
            pass

        def _link_builder_ptr(self: "Nop._Builder") -> LinkBuilderPtr:
            return LinkNop().link_nop()

        def _resolve_link(self: "Nop._Builder", _ptr: ControllerPtr) -> "Nop":
            return Nop(LinkNop().link_get(_ptr))

    def __init__(self: "Nop", ptr: LinkPtr) -> None:
        super().__init__(ptr)

    @staticmethod
    def builder() -> _Builder:
        """Create Nop link builder."""
        return Nop._Builder()

File: dotnet/cs/src/Link/Nop.cs
/*
 * File: Nop.cs
 * Project: Link
 * Created Date: 10/10/2023
 * Author: Shun Suzuki
 * -----
 * Last Modified: 29/11/2023
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2023 Shun Suzuki. All rights reserved.
 *
 */

using AUTD3Sharp.Internal;
using AUTD3Sharp.NativeMethods;

namespace AUTD3Sharp.Link
{
    /// <summary>
    /// Link which do nothing
    /// </summary>
    public class Nop
    {
        public sealed class NopBuilder : ILinkBuilder<Nop>
        {
            LinkBuilderPtr ILinkBuilder<Nop>.Ptr()
            {
                return NativeMethodsBase.AUTDLinkNop();
            }

            Nop ILinkBuilder<Nop>.ResolveLink(LinkPtr ptr)
            {
                return new Nop();
            }
        }

        public static NopBuilder Builder()
        {
            return new NopBuilder();
        }
    }
}
